dal/src/main/java/com/epam/esm/dal/impl/CertificateDaoSql.java

```diff
@@ -0,0 +1,154 @@
+		this.sqlBuilder = sqlBuilder;
+	}
+
+	@Transactional
+	@Override
+	public GiftCertificate addCertificate(GiftCertificate certificate) {
+		LocalDateTime lastUpdateDate = certificate.getLastUpdateDate();
+		Timestamp lastUpdateTimestamp = ((lastUpdateDate == null) ? null : Timestamp.valueOf(lastUpdateDate));
+		KeyHolder keyHolder = new GeneratedKeyHolder();
+		jdbcTemplate.update(new PreparedStatementCreator() {
+			@Override
+			public PreparedStatement createPreparedStatement(Connection con) throws SQLException {
+				String paramToReturn = "Id";
+				PreparedStatement preparedStatement = con.prepareStatement(sqlAddCertificate,
+						new String[] { paramToReturn });
+				preparedStatement.setString(1, certificate.getName());
+				preparedStatement.setString(2, certificate.getDescription());
+				preparedStatement.setDouble(3, certificate.getPrice());
+				preparedStatement.setTimestamp(4, Timestamp.valueOf(certificate.getCreationDate()));
+				preparedStatement.setTimestamp(5, lastUpdateTimestamp);
+				preparedStatement.setLong(6, certificate.getDuration());
+				return preparedStatement;
+			}
+		}, keyHolder);
+		long newSertificateId = keyHolder.getKey().longValue();
+		certificate.setId(newSertificateId);
+		updateTagsBoundedWithCertificate(certificate);
+		for (Tag tag : certificate.getTags()) {
+			jdbcTemplate.update(sqlInsertIntoM2M, tag.getId(), newSertificateId);
+		}
+		return certificate;
+	}
+
+	@Transactional
+	@Override
+	public GiftCertificate updateCertificate(GiftCertificate certificate) {
+		int affectedRows = jdbcTemplate.update(sqlUpdateCertificate, certificate.getName(),
+				certificate.getDescription(), certificate.getPrice(), certificate.getLastUpdateDate(),
+				certificate.getDuration(), certificate.getId());
+		if (certificate.getTags() != null && !certificate.getTags().isEmpty()) {
+			jdbcTemplate.update(sqlDeleteFromM2M, certificate.getId());
+			updateTagsBoundedWithCertificate(certificate);
+			for (Tag tag : certificate.getTags()) {
+				jdbcTemplate.update(sqlInsertIntoM2M, tag.getId(), certificate.getId());
+			}
+		}
+		if (affectedRows == 0) {
+			return null;
+		}
+		return certificate;
+	}
+
+	@Override
+	public List<GiftCertificate> findCertificates(List<FilterParam> filterParams, List<OrderParam> orderParams) {
+		String sqlQuery = sqlBuilder.buildCertificatesQuery(filterParams, orderParams);
+		try {
+			List<GiftCertificate> certificates = jdbcTemplate.query(sqlQuery, new CertificateWithTagsMapper());
+			certificates = DuplicateResultsRemover.removeDuplicateResults(certificates);
+			return certificates;
+		} catch (DataAccessException e) {
+			// nothing was found by the request;
+			return new ArrayList<GiftCertificate>();
+		}
+	}
+
+	@Override
+	public GiftCertificate findCertificate(long id) {
+		try {
+			return jdbcTemplate.queryForObject(sqlFindCertificateById, new Object[] { id },
+					new CertificateMapper());
+		} catch (DataAccessException e) {
+			return null;
```

Comment:
It's better to return Optional or throw an exception in such cases to avoid further NullPointers.